### ISSUE TITLE
Application random freezes

### ISSUE THREAD

**Issue by roberto-sartori** [xamarin/xamarin-android#6587]
### Android application type

Classic Xamarin.Android (MonoAndroid12.0, etc.)

### Affected platform version

Visual Studio 2019 (16.4.4) with Xamarin (16.4.000.309)

### Description

We have an application running on Android 7.1.1, developed with Xamarin.

This application is running 24h/24 on a device with screen always on. This application manages a lot of data, so it causes a lot of garbage collection events (one every 10 seconds, approximately).

Issue: sometimes, after some days, the application freezes without any error in the logs (logcat/dmesg). In this state, the app does not respond to touches or adb commands and no logs is printed anymore by the application.

Debugging, we saw that when the app freezes, the app process is in the _sigsuspend_ status (_ps_ command):
`system    15111 220   1877900 263016 sigsuspend b011d640 S com.android.app`

We then took a _strace_ of the process (attached) where we can see that the app enters the sigsuspend status here:
`--- SIGPWR {si_signo=SIGPWR, si_code=SI_TKILL, si_pid=1562, si_uid=1000} ---
rt_sigprocmask(SIG_BLOCK, [XCPU], [HUP INT QUIT ILL TRAP ABRT BUS FPE USR1 SEGV USR2 PIPE ALRM TERM STKFLT CHLD CONT TSTP TTIN TTOU URG XCPU XFSZ VTALRM PROF WINCH IO PWR SYS RTMIN], 8) = 0
futex(0xa5d0c5c4, FUTEX_WAKE_PRIVATE, 2147483647) = 1
rt_sigsuspend([HUP INT QUIT ILL TRAP ABRT BUS FPE KILL USR1 SEGV USR2 PIPE ALRM TERM STKFLT CHLD CONT STOP TSTP TTIN TTOU URG XFSZ VTALRM PROF WINCH IO PWR SYS RTMIN], 8
`

[strace_full.txt](https://github.com/xamarin/xamarin-android/files/7755639/strace_full.txt)

Basically, when the issue appears the _rt_sigsuspend_ function never returns.
It is my understanding that the _SIGPWR_ and _SIGXCPU_ signals are used by Mono to manage the garbage collections.
We tried with the tarjar implementation and with the old GC implementation: we have the issue with both, but using the _old_ implementation the GC events are rarer (one every 70 seconds).

### Steps to Reproduce

It's very difficult to replicate the issue, as it requires a specific application running on a specific hardware.

### Did you find any workaround?

The only way to get the application running again is to close it and reopen it, but this is not something feasible with our use case.

### Relevant log output

```shell
At every GC event:
01-17 04:00:59.386  1385  1427 I art     : Starting a blocking GC Explicit
01-17 04:00:59.401  1385  1419 I Mono    : GC_BRIDGE waiting for bridge processing to finish
01-17 04:00:59.438  1385  1427 I art     : Explicit concurrent mark sweep GC freed 13111(711KB) AllocSpace objects, 0(0B) LOS objects, 40% free, 3MB/5MB, paused 885us total 51.839ms
01-17 04:00:59.438  1385  1394 I Mono    : GC_BRIDGE waiting for bridge processing to finish
01-17 04:00:59.446  1385  1427 I Mono    : GC_TAR_BRIDGE bridges 457 objects 457 opaque 122 colors 457 colors-bridged 457 colors-visible 457 xref 61 cache-hit 0 cache-semihit 0 cache-miss 0 setup 0.22ms tarjan 0.51ms scc-setup 0.29ms gather-xref 0.04ms xref-setup 0.03ms cleanup 0.21ms
01-17 04:00:59.446  1385  1427 I Mono    : GC_BRIDGE: Complete, was running for 62.56ms
01-17 04:00:59.446  1385  1427 I Mono    : GC_MINOR: (Nursery full) time 13.00ms, stw 14.43ms promoted 85K major size: 6448K in use: 5109K los size: 4096K in use: 239K
```

**Comment by grendello** [comment]
This looks like a Mono runtime issue, @lambdageek can you take a look (or delegate), please? Thanks!

**Comment by lambdageek** [comment]
@roberto-sartori are you able to attach a native debugger (for example Android Studio) when the app hangs?  Can you get a stack trace?

The thread that hangs is stuck here:

https://github.com/mono/mono/blob/16b53fa04c9824cda209b35febdc51b4876fd1a0/mono/utils/mono-threads-posix-signals.c#L193-L200

but it is not clear what the rest of the app is doing.  

do you know if the app (or maybe some native library that the app depends on) installs a handler for `SIGXCPU` ?

Can you try setting the environment variable `MONO_SLEEP_ABORT_LIMIT` to some value like 2000 using `adb` before running the app. That will make the mono GC wait 2 seconds when trying to suspend and otherwise abort the app.  In that case we might get something in the adb logs with more information about what the app was doing.

**Comment by roberto-sartori** [comment]
@lambdageek 

> are you able to attach a native debugger (for example Android Studio) when the app hangs? Can you get a stack trace?

Can you clarify this? We have access to the application source code, but the code cannot be used in Android Studio (c#) so, even if I attach the debugger to the running app, I cannot insert breakpoints or actually debug the code from Android Studio. 

> do you know if the app (or maybe some native library that the app depends on) installs a handler for SIGXCPU ?

The application does not install a handler for SIGXCPU, neither do the native libraries from what we can see.

**Comment by jfichtner** [comment]
@roberto-sartori I believe @lambdageek is talking about the process described [here](https://github.com/xamarin/xamarin-android/blob/main/Documentation/workflow/DevelopmentTips.md). @lambdageek please advise.

**Comment by roberto-sartori** [comment]
> Can you try setting the environment variable MONO_SLEEP_ABORT_LIMIT to some value like 2000 using adb before running the app. That will make the mono GC wait 2 seconds when trying to suspend and otherwise abort the app. In that case we might get something in the adb logs with more information about what the app was doing.

We tried this, but we did not get any more logs. Same freeze and logs with and without the MONO_SLEEP_ABORT_LIMIT variable set.

@lambdageek can you provide more details about the information you need?
We setup Android Studio and attach its native debugger to the running app (built in debug) but we weren't getting useful logs.

I think this may be due to the fact that we need the _libmonosgen-2.0.so_ library with debug symbols, however following the documentation from the _Development tips and native debugging_ page did not get us to any download link. Can you check if the library is available to be downloaded?

Thank you.

**Comment by lambdageek** [comment]
> We tried this, but we did not get any more logs. Same freeze and logs with and without the MONO_SLEEP_ABORT_LIMIT variable set.

Ok, thanks for trying.

> 
> @lambdageek can you provide more details about the information you need? We setup Android Studio and attach its native debugger to the running app (built in debug) but we weren't getting useful logs.

It seems like Mono is hanging while trying to stop the user threads in order to run the GC.  I would like to see the native stack trace.  If you can attach LLDB or GDB to the running process, (and you have debugging symbols), `bt all` (lldb) or `thread apply all bt` (gdb) would give me something to work with.

> I think this may be due to the fact that we need the _libmonosgen-2.0.so_ library with debug symbols, however following the documentation from the _Development tips and native debugging_ page did not get us to any download link. Can you check if the library is available to be downloaded?

@jonpryor do you know if the links are upto date for pre-NET6 binary artifacts?

> Thank you.

**Comment by roberto-sartori** [comment]
Hi @lambdageek  , thank you for the information.
We have setup the debug session with Android Studio. We loaded the library with the debug symbols following [Option C: Manually load the library with symbols](https://github.com/xamarin/xamarin-android/blob/main/Documentation/workflow/DevelopmentTips.md).

We noticed, on different setups, that in this case the application freezes much more frequently (in some hours instead of some days). We don't think this is the issue we are actually trying to solve.
I attach anyway the log we get with the _bt all_ command from _lldb_. It appears that having the debugger attached causes the freeze for some reason.
It also appears that using Visual Studio we don't get this issue, after starting the debugging. 
Do you have any insight on this?

We have now different setups configured with Visual Studio, and we are going to get the logs using GDB when the actual freeze appears.
[android_studio_log.txt](https://github.com/xamarin/xamarin-android/files/8349408/android_studio_log.txt)

**Comment by roberto-sartori** [comment]
Hi @lambdageek ,
We were able to replicate the issue while debugging with Visual Studio.
Attached the output of the _thread apply all bt_ command.

Let me know if this is useful.
[vs_debug.txt](https://github.com/xamarin/xamarin-android/files/8511415/vs_debug.txt)

**Comment by jfichtner** [comment]
@lambdageek please see @roberto-sartori's post above, with debug output info. This issue is holding my team up from release of multiple application upgrades for multiple products that we have invested a lot of time and money into. On that note, I understand that most of the time so far has been waiting on us to replicate the issue and provide log output, but I was wondering if submitting an escalated support ticket for this issue might help you, or whomever else might be able to help with this issue, allocate more priority to it. Please let me know if this would help and I will create the ticket. Thanks!

**Comment by jfichtner** [comment]
@lambdageek I apologize, I'm sure you're busy, but I am quite stuck at the moment. Please see mine and Roberto's posts above. Thank you.

**Comment by lambdageek** [comment]
@BrzVlad Can you take a look at the stack trace in https://github.com/xamarin/xamarin-android/issues/6587#issuecomment-1102322375 `vs_debug.txt`.

It looks like Thread 46 is doign a gC and it's waiting for the jobs to finish.
But Thread 13 is the only worker (is it normal for there to only be one worker?) and it's waiting for more work.

**Comment by BrzVlad** [comment]
The stack trace suggests that the the problem lies within the implementation of pthread api in android libc and our preemptive suspend. Apparently pthread synchronization primitives on android can take a global lock. So if we suspend a thread while it is about to call some pthread api, all future pthread mutexes are potentially broken. The GC internally uses locks so therefore it deadlocks.

@roberto-sartori Looking in the trace, the GC hangs as part of waiting for the concurrent sweeper. Disabling this functionality might fix the issue (set env var  `MONO_GC_PARAMS=no-concurrent-sweep`). The proper fix would be using cooperative suspend

@lambdageek Aren't we already using hybrid suspend on Android ? Why is thread 47 suspended randomly in pthread code ?

**Comment by lambdageek** [comment]
> @roberto-sartori Looking in the trace, the GC hangs as part of waiting for the concurrent sweeper. Disabling this functionality might fix the issue (set env var `MONO_GC_PARAMS=no-concurrent-sweep`). The proper fix would be using cooperative suspend

We haven't really productized coop suspend on non-watchOS.  I think `no-concurrent-sweep` is worth trying.

> 
> @lambdageek Aren't we already using hybrid suspend on Android ? 

This is pre-NET6 Xamarin.Android, it's using preemptive suspend.

**Comment by jfichtner** [comment]
> This is pre-NET6 Xamarin.Android, it's using preemptive suspend.
@lambdageek is it possible to migrate our apps to NET6 at this point without MAUI? I've searched for answers to this question, but it's not clear to me.

**Comment by lambdageek** [comment]
> is it possible to migrate our apps to NET6 at this point without MAUI

@jfichtner The official guidance is that MAUI is supported way to build mobile applications using .NET 6.

However MAUI is built atop the .NET for Android SDK, so if you make an SDK-style project using the `net6.0-android` runtime identifier and reference the appropriate nugets, you will get an Android APK.  See, for example https://github.com/dotnet/maui-samples/tree/main/archive/HelloAndroid

Note that moving from pre-.NET6 Xamarin.Android to the .NET 6 Android SDK has all the same issues as moving from .NET Framework 4.x to .NET Core 3.x (or to .NET 5 or 6) on desktop - if you use AppDomains, or other legacy APIs they may not be available.

**Comment by roberto-sartori** [comment]
Hi @BrzVlad , @lambdageek .
I added the _MONO_GC_PARAMS=no-concurrent-sweep_ config to my setup.
I confirmed this from adb:
```
:/ $ logcat|grep -i "monodroid: Setting environment variable"
monodroid: Setting environment variable 'MONO_GC_PARAMS' to 'no-concurrent-sweep'
```

However, I see from the logcat:
```
01-23 20:55:16.997  3310  3402 I art     : Starting a blocking GC Explicit
01-23 20:55:17.048  3310  3402 I art     : Explicit concurrent mark sweep GC freed 12947(710KB) AllocSpace objects, 0(0B) LOS objects, 40% free, 3MB/6MB, paused 885us total 50.795ms
```

Is this expected? I'm referring in particular to the _Explicit concurrent mark sweep_ details.

Thanks.

**Comment by BrzVlad** [comment]
@roberto-sartori That is not our log, it is the java GC. Our logs are prefixed by `Mono`. The setting should be fine.

**Comment by jfichtner** [comment]
@BrzVlad , @lambdageek my testing so far has been positive in that I have not seen this freeze on the 3 instances that I have had running for 7 days now, which is very encouraging! So I would like to understand the tradeoffs of setting the MONO_GC_PARAMS to "no-concurrent-sweep". Is this equivalent to unchecking the "Use concurrent garbage collector" box in the Android Options UI for the Xamarin.Android project? If not, can you explain what exactly this setting does, and if there could be any negative effects on my application?

>  The proper fix would be using cooperative suspend

Just so that I'm clear, this would be a change in Mono, which has already been implemented in .NET 6, is that correct? So if setting the "no-concurrent-sweep" flag fixes my issue, then once I migrate to .NET 6 I shouldn't have to set that flag anymore?

Thanks again for this great work guys!

**Comment by BrzVlad** [comment]
@jfichtner Concurrent garbage collector refers to the marking phase (which is the longest part of a garbage collection). In this mode, the major collection does a short pause where it scans the roots, then resumes the application while the concurrent mark traverses the entire heap. Once the concurrent-mark is finished it will pause again to finish the collection. As part of the collection finish we need to reset all the major blocks for future allocation. This is done in the sweep phase. This phase can also be done concurrently or as part of the finishing pause. Using the `no-concurrent-sweep` option will increase the finishing pause of the major collections. This addition should be a fraction of the finish pause, but for larger heaps it can take a few tens of milliseconds, as a rough estimate. On .NET6 there should be no need to use this flag.

**Comment by jfichtner** [comment]
I just wanted to report that we have concluded that this is a proper workaround for the application freeze issue we were seeing. Thanks again to everyone for your help!

**Comment by jpobst** [comment]
With support for Classic Xamarin.Android ended [May 1st, 2024](https://dotnet.microsoft.com/en-us/platform/support/policy/xamarin), this issue is likely no longer relevant.

If this still persists in .NET 8+, please open a new issue with updated information based on `net8.0-android` or greater.  Please include a link to this issue for context.
